### PR TITLE
ENH PasswordExpirationMiddleware implementation

### PR DESCRIPTION
**Pull request by dnsl48**
The new middleware forcing users with expired password to change it.

Tested with
 - default LoginHandler
 - CMSLoginHandler
 - MFA (TOTP)

### CODE REVIEW

File: src/Security/PasswordExpirationMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+     *
+     * @return string
+     */
+    protected static function absoluteUrl($url): string
+    {
+        if (substr($url, 0, 1) === '/') {
+            // add BASE_URL explicitly if not absolute
+            $url = Controller::join_links(Director::absoluteBaseURL(), $url);
+        } else {
+            $url = Director::absoluteURL($url);
```
[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Security/PasswordExpirationMiddleware.php]: Actually, why can't you just use `Director::absoluteURL()` for this? It should handle the "is root relative" logic internally

File: tests/php/Security/PasswordExpirationMiddlewareTest.php [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+            PasswordExpirationMiddleware::config()->get('default_redirect');
+        $this->rollbackWhitelistedUrls =
+            PasswordExpirationMiddleware::config()->get('whitelisted_urls');
+        $this->rollbackWhitelistedUrlStartswith =
+            PasswordExpirationMiddleware::config()->get('whitelisted_url_startswith');
+
+        PasswordExpirationMiddleware::config()->set('enable', true);
+        PasswordExpirationMiddleware::config()->set('default_redirect', null);
+        PasswordExpirationMiddleware::config()->set('whitelisted_urls', []);
+        PasswordExpirationMiddleware::config()->set('whitelisted_url_startswith', []);
```
Comment (robbieaverill) [inline tests/php/Security/PasswordExpirationMiddlewareTest.php]: You're not calling `parent::setUp()` here, that should cause this test to fail - same with tearDown

File: tests/php/Security/PasswordExpirationMiddlewareTest.php [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+            $this->rollbackDefaultRedirect
+        );
+        PasswordExpirationMiddleware::config()->set(
+            'whitelisted_urls',
+            $this->rollbackWhitelistedUrls
+        );
+        PasswordExpirationMiddleware::config()->set(
+            'whitelisted_url_startswith',
+            $this->rollbackWhitelistedUrlStartswith
+        );
```
Comment (robbieaverill) [inline tests/php/Security/PasswordExpirationMiddlewareTest.php]: When you call `parent::setUp()` the config layer is automatically nested, so you don't need to reset any configuration values afterwards (including the current user)

File: src/Security/PasswordExpirationMiddleware.php [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    ];
+
+    /**
+     * Identify the current user from the request
+     *
+     * @param HTTPRequest $request
+     * @param callable $delegate
+     * @return HTTPResponse
+     */
+    public function process(HTTPRequest $request, callable $delegate)
```
Comment (robbieaverill) [inline src/Security/PasswordExpirationMiddleware.php]: You _could_ add an HTTPResponse type hint here. If you do your tests will fail, because the delegates you have in the tests don't return HTTPResponses

File: tests/php/Security/PasswordExpirationMiddlewareTest.php [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+        $response = $a->process($request, static function () use (&$executed, &$activeMember) {
+            $executed = true;
+            $activeMember = Security::getCurrentUser();
+        });
+
+        $this->assertFalse($executed);
+        $this->assertNull($activeMember);
+
+        $this->assertEquals(302, $response->getStatusCode());
+        $this->assertEquals(Director::absoluteURL('redirect-address-custom'), $response->getHeader('Location'));
```
Comment (robbieaverill) [inline tests/php/Security/PasswordExpirationMiddlewareTest.php]: Are you able to split these into four separate test methods? You could use data providers, or just duplicate the middleware creation lines

File: docs/en/04_Changelogs/4.5.0.md [hunk 6]
```diff
@@ -30,7 +30,31 @@ composer require silverstripe/login-forms
...
+If you don't want the middleware to be activate (default) you can disable it in your application config like that:
+
+```yml
+---
+Name: disable-passwordExpirationMiddleware
+After:
+  - '#coresecurity'
+---
+SilverStripe\Security\PasswordExpirationMiddleware:
+  enable: false
```
[the remark this answers is not in the document]
Comment (dnsl48) [reply docs/en/04_Changelogs/4.5.0.md]: What's the usual way for removing a preconfigured middleware?

[the remark this answers is not in the document]
Comment (robbieaverill) [reply docs/en/04_Changelogs/4.5.0.md]: You can set the middleware class definition to null, e.g.

```yaml
SilverStripe\Core\Injector\Injector:
  SilverStripe\Control\Director:
    properties:
      Middlewares:
        PasswordExpirationMiddleware: null # disables it by removing it from the list of dependencies
```

[the remark this answers is not in the document]
Comment (dnsl48) [reply docs/en/04_Changelogs/4.5.0.md]: Agree, that's definitely a better way. Pity it's bulkier and a bit harder to follow.

File: src/Security/PasswordExpirationMiddleware.php [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+     */
+    private static $enable = true;
+
+    /**
+     * List of URLs allowed to visit by users with password expired
+     * These URLs won't be redirected to the password change form.
+     * Requests to these URLs will have the Member NOT deauthenticated
+     * in case there is no password change form registered.
+     *
+     * @var array<string>
```
[the remark this answers is not in the document]
Comment (dnsl48) [reply src/Security/PasswordExpirationMiddleware.php]: Right. I think a more correct would be `array<int, string>` though?

[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Security/PasswordExpirationMiddleware.php]: AFAIK that's not valid PHPDoc syntax

[the remark this answers is not in the document]
Comment (dnsl48) [reply src/Security/PasswordExpirationMiddleware.php]: I guess not. However, Sam was particularly chasing that in [his PR](https://github.com/silverstripe/silverstripe-framework/pull/9103/files#diff-b50d5a291fb439ef0b0bcd40f82c6388L22) so we can do static analysis. I'm guessing PHPDoc may still be able to cope with it?

File: src/Security/PasswordExpirationMiddleware.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (dnsl48) [reply src/Security/PasswordExpirationMiddleware.php]: > why can't you just use Director::absoluteURL()

Because `absoluteURL` would return incorrect results for `SS_BASE_URL` which has non-root base.
E.g.
```php
Environment::set('SS_BASE_URL', 'http://localhost/non-root/');

$a = Director::absoluteURL('/Secret/');

$a === 'http://localhost/Secret/'; // we want http://localhost/non-root/Secret/
```

File: src/Security/PasswordExpirationMiddleware.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Security/PasswordExpirationMiddleware.php]: In this instance `string[]` is all you need. If we want to start adopting an experimental new type declaration system we should do that through an RFC - [it was removed from PSR-5](https://github.com/php-fig/fig-standards/pull/1102) recently, so it's not currently part of the spec at all =/

File: src/Security/PasswordExpirationMiddleware.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Security/PasswordExpirationMiddleware.php]: I realise you'd been discussing this with @tractorcow recently, but that sounds like a bug to me and could be addressed in a separate issue

File: src/Security/PasswordExpirationMiddleware.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (dnsl48) [reply src/Security/PasswordExpirationMiddleware.php]: yep, fair enough

File: src/Security/PasswordExpirationMiddleware.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (dnsl48) [reply src/Security/PasswordExpirationMiddleware.php]: That's not really a bug per se, but rather `Director::absoluteURL` doesn't cover the particular use case we have right here.

We already have a separate issue for discussing that: https://github.com/silverstripe/silverstripe-framework/issues/9111

File: src/Security/PasswordExpirationMiddleware.php (hunk 4 above)
Comment (dnsl48) [reply src/Security/PasswordExpirationMiddleware.php]: that's the definition from the `HTTPMiddleware` interface. I'll just remove the doc

Comment (robbieaverill) [reply src/Security/PasswordExpirationMiddleware.php]: Yeah but you can add return types in subclasses if you want to, it was just a suggestion - keep the doc block 😄

Comment (dnsl48) [reply src/Security/PasswordExpirationMiddleware.php]: Yeah, as you mentioned yourself, we cannot guarantee everything actually returns HTTPResponse, so that may actually be a breaking change :)

Comment (robbieaverill) [reply src/Security/PasswordExpirationMiddleware.php]: lol - touche

File: src/Security/PasswordExpirationMiddleware.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (dnsl48) [reply src/Security/PasswordExpirationMiddleware.php]: > Strict typing wise, this method can return false

yeah, I was tricked by the Director::absoluteURL docbloc :smile:

File: src/Security/MemberAuthenticator/LoginHandler.php [hunk 8]
```diff
@@ -263,7 +264,14 @@ protected function redirectToChangePassword()
...
         );
         $changedPasswordLink = Security::singleton()->Link('changepassword');
+        $changePasswordUrl = $this->addBackURLParam($changedPasswordLink);
+        $whitelistedUrl = Controller::join_links($changedPasswordLink, 'ChangePasswordForm');
 
-        return $this->redirect($this->addBackURLParam($changedPasswordLink));
+        $session = $this->getRequest()->getSession();
+        $passwordExpirationMiddleware = Injector::inst()->get(PasswordExpirationMiddleware::class);
+        $passwordExpirationMiddleware->setRedirect($session, $changePasswordUrl);
+        $passwordExpirationMiddleware->whitelistUrl($session, $whitelistedUrl);
```
Comment (robbieaverill) [inline src/Security/MemberAuthenticator/LoginHandler.php]: I'm still not convinced we really need both the explicit URL and the partially matched URL - if you just use partial matching, you can probably remove all of this coupling, right? Then you just use the config options for what the changepassword URL is, and what the allowed URLs can start with

Comment (dnsl48) [reply src/Security/MemberAuthenticator/LoginHandler.php]: yeah, I guess that might be superfluous in our circumstances. Will remove in sake of reducing coupling and introduced API surface.

Comment (dnsl48) [reply src/Security/MemberAuthenticator/LoginHandler.php]: done

File: docs/en/04_Changelogs/4.5.0.md [hunk 9]
```diff
@@ -30,7 +30,34 @@ composer require silverstripe/login-forms
...
+A new PasswordExpirationMiddleware has been implemented.
+It checks passwords of authenticated users for expiration and either enforces a redirection
+to a change password form, or resets the user for a request being processed (sets current user to null).
+
+This is considered to be a security enhancement, but potentially might interfere with some custom logic
+around password expiration if you have it implemented.
+
+Ideally you should test your setup when upgrading if you use the password expiration functionality.
+
+If you'd like to deactivate the middleware, you can unregister it in your application config like that:
```
Comment (robbieaverill) [inline docs/en/04_Changelogs/4.5.0.md]:
```suggestion
If you'd like to deactivate the middleware, you can unregister it in your application config like this:
```

File: src/Security/PasswordExpirationMiddleware.php [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+     *
+     * @config
+     */
+    private static $default_redirect = null;
+
+    /**
+     * Whether we should allow the current request to be finished as is even if
+     * the user password has been expired.
+     *
+     * WARNING: this is not a configurable attribute, but should be controlled by a LoginHandler
```
Comment (robbieaverill) [inline src/Security/PasswordExpirationMiddleware.php]: Maybe it could be an instance property with a setter instead then

File: src/Security/PasswordExpirationMiddleware.php [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+     * Builds an absolute URL for the given path, adds base url
+     * if the path configured as absolute
+     *
+     * @param string $url
+     *
+     * @return string
+     */
+    protected static function absoluteUrl($url): string
+    {
+        if (substr($url, 0, 1) === '/') {
```
Comment (robbieaverill) [inline src/Security/PasswordExpirationMiddleware.php]: Just checking, we are now saying that `//localhost/foo` is not absolute, right? This check would return true for that URL

Comment (tractorcow) [reply src/Security/PasswordExpirationMiddleware.php]: > Just checking, we are now saying that //localhost/foo is not absolute, right? This check would return true for that URL

Not until ss5 at the earliest.

File: src/Security/PasswordExpirationMiddleware.php [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+        $currentUrl = $this->absoluteUrl($request->getURL(true));
+
+        $allowedUrls = array_merge([
+            $redirectUrl
+        ], array_map(
+            function ($url) {
+                return $this->absoluteUrl($url);
+            },
+            $whitelist
```
Comment (tractorcow) [inline src/Security/PasswordExpirationMiddleware.php]: array_map is being called on an empty array here isn't it?

File: src/Security/PasswordExpirationMiddleware.php (hunk 11 above)
Comment (tractorcow) [reply src/Security/PasswordExpirationMiddleware.php]: I've explained at https://github.com/silverstripe/silverstripe-framework/issues/9111#issuecomment-527009973 why you should not be treating root-relative urls as though they were base-relative. If you have a URL with a leading slash that you don't want to treat as a root-relative url, your data being put into the method is wrong.

Comment (dnsl48) [reply src/Security/PasswordExpirationMiddleware.php]: > Just checking, we are now saying that //localhost/foo is not absolute, right?

Technically, it's not an absolute URL, but it is still either an `absolute path` or a `network location` (relative URL starting with //). This check is basically for paths, not urls.

This is a good catch, by the way. Even though it's an edge case, in case of a network location this would produce incorrect result. I'll patch it up and cover with a test.

File: src/Security/PasswordExpirationMiddleware.php (hunk 12 above)
Comment (dnsl48) [reply src/Security/PasswordExpirationMiddleware.php]: yes, a leftover after some deleted functionality. Will fix it :+1:

File: _config/security.yml [hunk 13]
```diff
@@ -42,3 +43,11 @@ SilverStripe\Core\Injector\Injector:
       Authenticators:
         cms: '%$SilverStripe\Security\MemberAuthenticator\CMSMemberAuthenticator'
   SilverStripe\Security\IdentityStore: '%$SilverStripe\Security\AuthenticationHandler'
+
+SilverStripe\Security\PasswordExpirationMiddleware:
+  default_redirect: /Security/changepassword
```
Comment (tractorcow) [inline _config/security.yml]: I think removing the slash from the start of these urls should let you use the absoluteLink() method the way you expect, right?

File: src/Security/MemberAuthenticator/LoginHandler.php [hunk 14]
```diff
@@ -263,7 +264,13 @@ protected function redirectToChangePassword()
             'good'
         );
         $changedPasswordLink = Security::singleton()->Link('changepassword');
+        $changePasswordUrl = $this->addBackURLParam($changedPasswordLink);
+        $whitelistedUrl = Controller::join_links($changedPasswordLink, 'ChangePasswordForm');
 
-        return $this->redirect($this->addBackURLParam($changedPasswordLink));
+        $session = $this->getRequest()->getSession();
+        $passwordExpirationMiddleware = Injector::inst()->get(PasswordExpirationMiddleware::class);
+        $passwordExpirationMiddleware->setRedirect($session, $changePasswordUrl);
```
Comment (unclecheese) [inline src/Security/MemberAuthenticator/LoginHandler.php]: This block of code is exactly the same as in the subclass, only without the `allowCurrentRequest`. Do you think `LoginHandler` should have a `getPasswordExpirationMiddleware` function to consolidate this?

File: src/Security/PasswordExpirationMiddleware.php [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+     * Builds an absolute URL for the given path, adds base url
+     * if the path configured as absolute
+     *
+     * @param string $url
+     *
+     * @return string
+     */
+    protected static function absoluteUrl($url): string
+    {
+        if (substr($url, 0, 1) === '/' && substr($url, 1, 1) !== '/') {
```
Comment (unclecheese) [inline src/Security/PasswordExpirationMiddleware.php]: You can use square bracket syntax for this.. `$url[0] === '/' && $url[1] !== '/'`

File: src/Security/MemberAuthenticator/LoginHandler.php (hunk 14 above)
Comment (robbieaverill) [reply src/Security/MemberAuthenticator/LoginHandler.php]: I still think you can get rid of this code, and let it rely on config

File: src/Security/PasswordExpirationMiddleware.php (hunk 15 above)
Comment (dnsl48) [reply src/Security/PasswordExpirationMiddleware.php]: Then we would have to add two extra checks for `strlen($url)`, otherwise it may produce notices for empty strings and with one character length.

File: _config/security.yml (hunk 13 above)
Comment (dnsl48) [reply _config/security.yml]: I disagree. This would be semantically incorrect.

According to [RFC1808, Relative Uniform Resource Locators](https://tools.ietf.org/html/rfc1808#section-2.2) lack of leading slash would indicate the redirect should be relative to the request URI, which should not be the case in this particular case.

As such, `Director::absoluteURL` method in its current implementation is not appropriate for the current scenario.

File: src/Security/MemberAuthenticator/LoginHandler.php (hunk 14 above)
Comment (dnsl48) [reply src/Security/MemberAuthenticator/LoginHandler.php]: > I still think you can get rid of this code, and let it rely on config

Yeah, unfortunately we cannot completely get rid of it and keep CMSLoginHandler working. However, we might get rid of it in LoginHandler though. I'll have a look if we can redo this keeping in mind that we want to reduce coupling between login handlers and the new middleware...

Comment (dnsl48) [reply src/Security/MemberAuthenticator/LoginHandler.php]: Done. We still have to keep it in CMSLoginHandler though.

Comment (dnsl48) [reply src/Security/MemberAuthenticator/LoginHandler.php]: Ok, it broke something. Looking into it

Comment (dnsl48) [reply src/Security/MemberAuthenticator/LoginHandler.php]: Fixed, added a new test

File: _config/security.yml (hunk 13 above)
Comment (tractorcow) [reply _config/security.yml]: I actually read that RFC and it completely backs up what I have been trying to tell you.

image

>  (3.1) Base URL embedded in the document's content 

This is Director::baseURL().

Please remove the leading slashes and the custom url resolution logic.

Comment (dnsl48) [reply _config/security.yml]: > (3.1) Base URL embedded in the document's content

Yes, agree, however that's not what I'm talking about.

1. Middlewares work out of context of documents, since routing hasn't been done yet. There is no document identified for serving yet, so `3.3` applies, and not `3.1`.

> (3.3) URL used to retrieve the entity

If we ignore that and say that `SS_BASE_URL` must be applied everywhere and for everything in the framework, then we cannot remove the custom logic, because Director::absoluteURL doesn't follow that rule.

Comment (dnsl48) [reply _config/security.yml]: > Director::absoluteURL doesn't follow that rule

To elaborate - simply using it would produce incorrect results for absolute paths, which are relative URLs. As long as we don't have any rules about prohibiting use of leading slashes in the config files, I think that would be a wrong decision to make, leading to bugs in users projects.

Comment (unclecheese) [reply _config/security.yml]: Just had a read through both points here, and I'm leaning toward what @dnsl48 is saying. To summarise:

When given a leading slash, `Director::absoluteURL()` becomes a framework-agnostic utility function for generating an absolute URL against the current request. This completely subverts the `SS_BASE_URL` setting, which is trouble for those who are:

* Using a canonical domain name
* Running their project in a subdirectory

Given that, I don't think 3.1 of the RFC applies here, because no base URL context provided, in the document, nor in the middleware. Further, the two above conditions make the `Director::absoluteURL()` function unreliable in some contexts.

Briefly, if we use `Director::absoluteURL()`, we open up a window for errors and unexpected behaviour, and if we don't, we may have some redundant code, depending on how you interpret an RFC. I prefer the latter.

Comment (tractorcow) [reply _config/security.yml]: > When given a leading slash, Director::absoluteURL() becomes a framework-agnostic utility function for generating an absolute URL against the current request. This completely subverts the SS_BASE_URL setting, which is trouble for those who are:

This is why I instructed him to remove the leading slashes. Remove the leading slashes, and you have no problems. There's no question about how leading slashes should be handled, they are absolute path according to the RFC. Even @dnsl48 understood this.

The question is about how relative paths should be handled, which is the point we disagree with, hence his apprehension in using them.

>  Further, the two above conditions make the Director::absoluteURL() function unreliable in some contexts.

It's only unreliable to users who don't (or refuse) to understand.

Relative paths could be handled one of two ways (second argument to absoluteURL()):

 - relative to application base path (rfc behavioru in presence of a base url, preferred behaviour in 99% of times this method is called)
 - relative to current requets (rfc behaviour in absence of base url, but virtually never the desired outcome of devs calling this method).

Needless the say, the current 4.x default behaviour is to default in favour of the base path as the relative parent, not the current request. That's not a matter of opinion, it's a fact.

If this is to change, then that's another topic you are welcome to discuss in the context of 5.x, and potentially in the context of removing base_url entirely (something which many devs desire in the long term).

However, in the case of getting a new feature merged into the 4 branch of framework, you need to both understand and accept the current convention for this major feature. Changing an existing url resolution pattern is irrelevant in the context of getting this feature merged.

Comment (tractorcow) [reply _config/security.yml]: > If we ignore that and say that SS_BASE_URL must be applied everywhere and for everything in the framework, then we cannot remove the custom logic, because Director::absoluteURL doesn't follow that rule.

No, only for relative url resolution. Not "absolutely everything". You are intentionally misrepresenting what I said.

Comment (tractorcow) [reply _config/security.yml]: > To elaborate - simply using it would produce incorrect results for absolute paths, which are relative URLs. As long as we don't have any rules about prohibiting use of leading slashes in the config files, I think that would be a wrong decision to make, leading to bugs in users projects.

This is why I said not to use absolute paths. You surely read the first of my very easy change requests right?

Comment (dnsl48) [reply _config/security.yml]: > You are intentionally misrepresenting what I said.

Sorry if I did that, it wasn't intentional. I meant to say about it in a general way as an opposite to the other option and wasn't trying to refer to your words there.

> The question is about how relative paths should be handled, which is the point we disagree with

Sorry, but that's not where I'm disagreeing with you.
I rather consider `Remove the leading slashes, and you have no problems` `It's only unreliable to users who don't (or refuse) to understand` to be a sub-optimal solution.

Configuring the `default redirect path` for the middleware people won't know how exactly it's gonna be resolved. They don't have to know details of Director::absoluteURL implementation and most likely they won't be even knowing that PasswordExpirationMiddleware is using it internally.

Users are going to be putting the leading slash into the configs because it feels semantically right according to the web standards. By removing the custom logic around absolute url resolutions in this middleware we would introduce a surface for bugs occurring there.

I can remove the leading slash from the config, since it's closer to our current conventions (which aren't documented by the way, that's why I'm one of those who don't know about them). However, it feels like (and I've already said that I actually don't really know them) those conventions are rather about routing and not completely applicable to redirects. Routing doesn't really have to deal with how absolute vs relative paths should be handled. Redirects do.

I think Director::absoluteURL is a red herring here. I'd rather be thinking about how the middleware should work and what it should be doing with its configuration.
So far I've got tests covering both use cases - lack of the leading slash and its presence. As long as the conventions are not rules and we don't prohibit leading slashes, I believe it would be more flexible and less error prone to handle both gracefully rather than remove support for absolute paths and demand people not to use leading slashes.
